### PR TITLE
Feature: Set drag zone with debug toolbar area excluded

### LINKED ISSUE #11592: Feature: Support "Select Element" feature in running application
### What feature or improvement do you think would benefit Files?

UFX8LALZ{N67(}TT CKG4(6

When an app runs, we can see the in-app toolbar appear within the running application. It is useful when debugging and positioning bugs.

But because of the set of `AppWindowTitleBar.SetDragRectangles()`. The area of toolbar is blocked.

I hope when in debug mode, the area of toolbar can be made available for use.

### Requirements

- This proposal will support **Select Element** feature in running application (only available in DEBUG mode)

### Files Version

2.0.36.0

### Windows Version

Windows11 22H2 22621.1265

### Comments

_No response_

### PR DESCRIPTION
**Pull request by Poker-sang**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes  #11592 

**Details**

Toolbar is useful when debugging. So I subtract the toolbar area before calling `AppWindowTitleBar.SetDragRectangles()` , only when the app is running in DEBUG mode. Now the area of toolbar won't be blocked any longer.

Method `GetDragZone()` is an algorithm for rectangle subtraction, and it is well suited to expand other use. For example, excluding a new area for new avatar control.

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

### CODE REVIEW

File: src/Files.App/Views/MainPage.xaml.cs [hunk 1]
```diff
@@ -263,29 +255,29 @@ private void SidebarControl_SidebarItemInvoked(object sender, SidebarItemInvoked
 			switch ((invokedItemContainer.DataContext as INavigationControlItem)?.ItemType)
 			{
 				case NavigationControlItemType.Location:
-					{
```
Comment (yaira2) [inline src/Files.App/Views/MainPage.xaml.cs:266]: Were these supposed to be changed?

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		/// <param name="nonDraggingZones"></param>
+		public static void SetDragZones(Window window, int dragZoneHeight = 60, int dragZoneLeftIndent = 0, IEnumerable<RectInt32>? nonDraggingZones = null)
+		{
+			var hWnd = WindowNative.GetWindowHandle(window);
+			var windowId = Win32Interop.GetWindowIdFromWindow(hWnd);
+			var appWindow = AppWindow.GetFromWindowId(windowId);
+			var scaleAdjustment = GetScaleAdjustment(window);
+			nonDraggingZones ??= Array.Empty<RectInt32>();
+#if DEBUG
+			nonDraggingZones = nonDraggingZones.Concat(new RectInt32[] { new((appWindow.Size.Width - DebugToolbarWidth) / 2, 0, DebugToolbarWidth, DebugToolbarHeight) });
```
Comment (yaira2) [inline src/Files.App/Helpers/DragZoneHelper.cs:127]: I like this! Can you add a comment?

Comment (Poker-sang) [reply src/Files.App/Helpers/DragZoneHelper.cs]: Certainly, I will do it after school,

File: src/Files.App/Views/MainPage.xaml.cs (hunk 1 above)
Comment (Poker-sang) [reply src/Files.App/Views/MainPage.xaml.cs:266]: Sorry. It seems to be the result of code formatting. I will revert it later.

**Comment by Poker-sang** [comment]
I just push to fix bug about screen scaling. I found that `AppWindow.Size` is related to DPI, which `UIElement`s don't.

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+namespace Files.App.Helpers
+{
+	public static class DragZoneHelper
+	{
+		/// <summary>
+		/// Get Scale Adjustment
+		/// </summary>
+		/// <returns>scale factor percent</returns>
+		/// <param name="window"></param>
```
Comment (ferrariofilippo) [inline src/Files.App/Helpers/DragZoneHelper.cs:19]: Can you swap these two lines?

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+							draggingZonesX.RemoveAt(i);
+							draggingZonesY.RemoveAt(i);
+							if (xResult[0].Lower == x.Lower)
+							{
+								draggingZonesY.InsertRange(i, new[] { y, yResult });
+								draggingZonesX.InsertRange(i, new[]
+								{
+								x with { Upper = xResult[0].Upper },
+								x with { Lower = xSubtrahend.Lower }
+							});
```
Comment (ferrariofilippo) [inline src/Files.App/Helpers/DragZoneHelper.cs:63]: Spacing seems off

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+							}
+							else // xResult[0].Upper == x.Upper
+							{
+								draggingZonesY.InsertRange(i, new[] { yResult, y });
+								draggingZonesX.InsertRange(i, new[]
+								{
+								x with { Upper = xSubtrahend.Upper },
+								x with { Lower = xResult[0].Lower }
+							});
```
Comment (ferrariofilippo) [inline src/Files.App/Helpers/DragZoneHelper.cs:72]: As above

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+						case 2:
+							draggingZonesX.RemoveAt(i);
+							draggingZonesY.RemoveAt(i);
+							draggingZonesY.InsertRange(i, new[] { y, yResult, y });
+							draggingZonesX.InsertRange(i, new[]
+							{
+							x with { Upper = xResult[0].Upper } ,
+							xSubtrahend,
+							x with { Lower = xResult[1].Lower }
+						});
```
Comment (ferrariofilippo) [inline src/Files.App/Helpers/DragZoneHelper.cs:86]: As above

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+							++i;
+							++i;
```
Comment (ferrariofilippo) [inline src/Files.App/Helpers/DragZoneHelper.cs:88]: Wouldn't `i += 2` be better for readability and performance?

**Review by ferrariofilippo — CHANGES_REQUESTED** [review]
I think you should apply this also to Advanced Security Window (`Files.App.Views.Properties.SecurityPage.OpenSecurityAdvancedPageWindow()`)

File: src/Files.App/Helpers/DragZoneHelper.cs (hunk 4 above)
Comment (Poker-sang) [reply src/Files.App/Helpers/DragZoneHelper.cs:63]: Thanks

File: src/Files.App/Helpers/DragZoneHelper.cs (hunk 3 above)
Comment (Poker-sang) [reply src/Files.App/Helpers/DragZoneHelper.cs]: Ok

**Comment by Poker-sang** [comment]
> I think you should apply this also to Advanced Security Window (`Files.App.Views.Properties.SecurityPage.OpenSecurityAdvancedPageWindow()`)

Thank you. But I thought there is need for this, because when initializing, all `SizeChanged` event will be raised.

**Comment by ferrariofilippo** [comment]
That's a different window, it's not Properties one

**Comment by Poker-sang** [comment]
> That's a different window, it's not Properties one

Ok, I implemented it.

**Comment by ferrariofilippo** [comment]
I'm not sure why, but when I open `Properties` window, I have to manually resize the window to interact with the debug toolbar.
With `Advanced Security Window` and `Main Page` it works since I open them

**Comment by Poker-sang** [comment]
I will check it out later

**Comment by Poker-sang** [comment]
> I'm not sure why, but when I open `Properties` window, I have to manually resize the window to interact with the debug toolbar.
> With `Advanced Security Window` and `Main Page` it works since the beginning

It is odd. I couldn't figure it out either. Anyway I trigger it once programatically when it starts.

**Review by ferrariofilippo — DISMISSED** [review]
LGTM and Works 👍

File: src/Files.App/Helpers/DragZoneHelper.cs (hunk 2 above)
Comment (yaira2) [reply src/Files.App/Helpers/DragZoneHelper.cs]: 👍

File: src/Files.App/Helpers/DragZoneHelper.cs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+		/// <param name="nonDraggingZones"></param>
+		public static void SetDragZones(Window window, int dragZoneHeight = 40, int dragZoneLeftIndent = 0, IEnumerable<RectInt32>? nonDraggingZones = null)
+		{
+			var hWnd = WindowNative.GetWindowHandle(window);
+			var windowId = Win32Interop.GetWindowIdFromWindow(hWnd);
+			var appWindow = AppWindow.GetFromWindowId(windowId);
+			var scaleAdjustment = GetScaleAdjustment(window);
+			var windowWidth = (int)(appWindow.Size.Width / scaleAdjustment);
+			nonDraggingZones ??= Array.Empty<RectInt32>();
+#if DEBUG   // subtract the toolbar area (center-top in window), only in DEBUG mode.
```
Comment (yaira2) [inline src/Files.App/Helpers/DragZoneHelper.cs:127]: Might be a good idea to move the comment to a new line.

Comment (Poker-sang) [reply src/Files.App/Helpers/DragZoneHelper.cs]: Done

**Comment by yaira2** [comment]
@Poker-sang can you resolve the merge conflicts?

**Comment by Poker-sang** [comment]
@yair100 All right. Resolved

**Review by yair100 — APPROVED** [review]
LGTM

**Pull request merged**